cpp/include/cudf/detail/iterator.cuh

```diff
@@ -153,6 +154,53 @@ auto inline make_validity_iterator(column_device_view const& column)
       validity_accessor{column});
 }
 
+/**
+ * @brief Constructs a constant iterator over a scalar's value.
+ *
+ * Dereferencing the returned iterator returns a `Element`.
+ * 
+ * For `p = *(iter + i)`, `p` is the value stored in the scalar.
+ *
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ * @throws `cudf::logic_error` if scalar is null.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar value
+ */
+template <typename Element>
+auto inline make_scalar_iterator(scalar const& scalar_value)
+{
+  CUDF_EXPECTS(data_type(experimental::type_to_id<Element>()) == scalar_value.type(), "the data type mismatch");
+  CUDF_EXPECTS(scalar_value.is_valid(), "the scalar value must be valid");
+  using ScalarType = experimental::scalar_type_t<Element>;
+  return thrust::make_constant_iterator(static_cast<ScalarType const*>(&scalar_value)->value());
+}
+
+/**
+ * @brief Constructs a constant pair iterator over a scalar's value and its validity.
+ *
+ * Dereferencing the returned iterator returns a `thrust::pair<Element, bool>`.
+ *  
+ * If scalar is valid, then for `p = *(iter + i)`, `p.first` contains 
+ * the value of the scalar and `p.second == true`.
+ *
+ * Else, if the scalar is null, then the value of `p.first` is undefined and `p.second == false`. 
+ * 
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar, and validity of the scalar in a pair
+ */
+template <typename Element>
+auto inline make_pair_iterator(scalar const& scalar_value) {
```

Comment:
I really need a make_pair_iterator() that takes either a column_view or a column_device_view and also returns the same thing as this scalar version  (thrust::pair<Element, bool>).

The overall idea is to be able to access a column's values and validity vector in a generic way while also being able to treat a scalar as a virtual column,  so that functions like these can operated on in a generic way.

do_some_work(column_view left, column_view right)
do_some_work(column_view left, scalar right)
do some_work(scalar left, column_view right)

Reply:
This PR has taken ages and it is blocking other PRs merging. Please file a separate issue for this and/or create a PR for what you need.

Reply:
@nvdbaranec I will bring back `make_pair_iterator(column_device_view)`. For  column_view, please create a separate issue.

Reply:
@nvdbaranec do you really need make_pair_iterator(column_device_view), or would you be OK with `column_device_view::pair_begin()` and `pair_end()`? I prefer the latter.

Reply:
Preference would be make_pair_iterator.  The context here is the refactor of copy_if_else() / scalars.  Specifically, the cleanest refactoring I can do here: https://github.com/rapidsai/cudf/pull/3471#pullrequestreview-323362862

involves wrapping the various inputs (column/column, scalar/column, column/scalar, scalar/scalar) at the highest possible level with iterators and sending those down to the detail implementation.   There's one templated function between the public implementation and the detail layer that takes generic Left and Right types which can be either scalars or column_views - being able to just pass either type to make_pair_iterator() would be pretty clean.